### PR TITLE
Feat/maintenance report

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/860rv3gg5

### :video_camera: Screenshots/Screen capture

https://github.com/EyeSeeTea/d2-reports/assets/37223065/30558258-65fe-49ab-a8f2-4e1693738703

### :fire: Notes to the tester
```
REACT_APP_DHIS2_BASE_URL=https://dev.eyeseetea.com/who-dev-238/
REACT_APP_REPORT_VARIANT=glass-admin
```

### CODE REVIEW

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        const userModules = modules.filter(module => {
+            const moduleUserGroupIds = module.userGroups.approveAccess?.map(userGroup => userGroup.id) ?? [];
+
+            return _.some(moduleUserGroupIds, moduleUserGroupId => userGroupIds.includes(moduleUserGroupId));
+        });
+
+        return userModules;
+    }
+
+    async delete(namespace: string, items: Id[]): Promise<void> {
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:66]: items -> itemIds. But item is too generic, what is the entity we are working with?

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        return this.storageClient.saveObject<string[]>(namespace, columns);
+    }
+
+    private async getCountries(): Promise<NamedRef[]> {
+        const { organisationUnits } = await this.api
+            .get<{ organisationUnits: NamedRef[] }>("/organisationUnits", {
+                fields: "id, name",
+                filter: "level:eq:3",
+                paging: false,
+            })
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:90]: We should use `this.api.models.organisationUnits` or `this.api.metadata.get({organisationUnits: ...})`

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    private async getModules(): Promise<GLASSModule[]> {
+        const modules =
+            (await this.globalStorageClient.getObject<GLASSModule[]>(Namespaces.DATA_SUBMISSSIONS_MODULES)) ?? [];
+
+        return modules
+            .map(module => ({
+                id: module.id,
+                name: module.name,
+                userGroups: module.userGroups,
+            }))
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:105]: All is the same, no need to map.

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                console.debug("File does not exist");
+            }
+        });
+    }
+
+    private getFilteredFiles(files: GLASSDataMaintenanceItem[], orgUnits: NamedRef[], module: Module) {
+        const lastDataSubmissionYear = new Date().getFullYear() - 1; // last submission year is the previous year
+
+        return _(files)
+            .map(upload => ({
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:129]: be consistent: files/upload, should be the same, one plural, the other singular.

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    private updateStatusById(
+        ids: string[],
+        uploads: GLASSDataMaintenanceItem[],
+        status: Status
+    ): GLASSDataMaintenanceItem[] {
+        return _.map(uploads, upload => {
+            if (_.includes(ids, upload.fileId)) {
+                return _.assign({}, upload, { status: status });
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:151]: `{...upload, {status: status}}`

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        };
+    }
+
+    private getPaginatedObjects<T>(rows: T[], sorting: Sorting<T>, paging: Paging): T[] {
+        return _(rows)
+            .orderBy([row => row[sorting.field]], [sorting.direction])
+            .drop((paging.page - 1) * paging.pageSize)
+            .take(paging.pageSize)
+            .value();
+    }
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:179]: Implement logic in entity: PaginatedObjects

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    private paginate<T>(objects: T[], sorting: Sorting<T>, paging: Paging) {
+        const pager = this.getPager(objects, paging);
+        const paginatedObjects = this.getPaginatedObjects(objects, sorting, paging);
+
+        return { pager, objects: paginatedObjects };
+    }
+}
+
+const emptyPage = {
```
Comment (tokland) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:189]: When definining literal objects, be explicit with the type so TS can help you: `: GLASSMaintenancePaginatedObjects<GLASSDataMaintenanceItem>`

File: src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    orgUnit: string;
+    orgUnitName: string;
+    period: string;
+    status: Status;
+}
+
+export interface GLASSModule extends NamedRef {
+    name: Module;
+    userGroups: {
+        approveAccess?: NamedRef[];
```
Comment (tokland) [inline src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx:22]: As a general rule, minimize optional properties. I guess this comes from data not being set in the data store. Then the solution is setting default values when loading from the data store. 

That said, if unsure, keep it as is.

File: src/domain/reports/glass-admin/repositories/GLASSDataMaintenanceRepository.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    GLASSDataMaintenanceItem,
+    GLASSMaintenancePaginatedObjects,
+    GLASSModule,
+    Module,
+} from "../entities/GLASSDataMaintenanceItem";
+
+export interface GLASSDataMaintenanceOptions {
+    paging: Paging;
+    sorting: Sorting<GLASSDataMaintenanceItem>;
+    module: Module | "";
```
Comment (tokland) [inline src/domain/reports/glass-admin/repositories/GLASSDataMaintenanceRepository.ts:14]: As a general rule, `| ""` looks weird. Typically you'd write `Module | undefined` (not sure in this particular scenario)

File: src/domain/reports/glass-admin/usecases/UpdateGLASSDataMaintenanceUseCase.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { UseCase } from "../../../../compositionRoot";
+import { GLASSDataMaintenanceRepository } from "../repositories/GLASSDataMaintenanceRepository";
+
+export class UpdateGLASSDataMaintenanceUseCase implements UseCase {
```
Comment (tokland) [inline src/domain/reports/glass-admin/usecases/UpdateGLASSDataMaintenanceUseCase.ts:4]: We used to do this "implements UseCase", but we have noticied it hinders code navigation while adding very little value, so better just to remove it.

File: src/domain/reports/glass-admin/usecases/UpdateGLASSDataMaintenanceUseCase.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    constructor(private maintenanceRepository: GLASSDataMaintenanceRepository) {}
+
+    execute(namespace: string, action: UpdateAction, items: string[]): Promise<void> | undefined {
+        switch (action) {
+            case "delete":
```
Comment (tokland) [inline src/domain/reports/glass-admin/usecases/UpdateGLASSDataMaintenanceUseCase.ts:9]: No need to change it, but typically we create separate use cases, avoid catch-all use cases.

File: src/webapp/reports/glass-admin/glass-admin-list/DataMaintenanceList.tsx [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+    return (
+        <React.Fragment>
+            <Filters values={filters} onChange={setFilters} />
+
+            <TabHeader labels={reportTabs} tabIndex={tabIndex} onChange={handleChange} />
+
+            <TabPanel value={tabIndex} index={0}>
+                <StyledButtonContainer>
+                    <Button onClick={async () => deleteFiles(filesToDelete)} color="primary" variant="contained">
```
Comment (tokland) [inline src/webapp/reports/glass-admin/glass-admin-list/DataMaintenanceList.tsx:88]: No need for async. Move to useCallback

File: src/webapp/reports/glass-admin/glass-admin-list/useFiles.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            await compositionRoot.glassAdmin.saveColumns(Namespaces.FILE_UPLOADS_USER_COLUMNS, columnKeys);
+        },
+        [compositionRoot.glassAdmin, visibleColumns]
+    );
+
+    async function deleteFiles(ids: string[]) {
+        openDeleteModal();
+        compositionRoot.glassAdmin.updateStatus(Namespaces.FILE_UPLOADS, "delete", ids)?.then(() => closeDeleteModal());
+        reload();
+    }
```
Comment (tokland) [inline src/webapp/reports/glass-admin/glass-admin-list/useFiles.tsx:69]: Move all these useMemo / useCallback to custom hooks. This reduces lines count and make the render block more declarative.

File: src/webapp/reports/glass-admin/glass-admin-list/useFiles.tsx [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    const [isDeleteModalOpen, { enable: openDeleteModal, disable: closeDeleteModal }] = useBooleanState(false);
+
+    useEffect(() => {
+        compositionRoot.glassAdmin.getColumns(Namespaces.FILE_UPLOADS_USER_COLUMNS).then(columns => {
+            setVisibleColumns(columns);
+        });
+    }, [compositionRoot.glassAdmin]);
+
+    const getFiles = useMemo(
+        () => async (_search: string, paging: TablePagination, sorting: TableSorting<DataMaintenanceViewModel>) => {
```
Comment (tokland) [inline src/webapp/reports/glass-admin/glass-admin-list/useFiles.tsx:38]: A useMemo of a function is typically a useCallback.

**Review by tokland — COMMENTED** [review]
Looking good, some in-line comments, mostly minor issues.

File: src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx [hunk 15]
```diff
@@ -19,10 +21,23 @@ export interface GLASSDataMaintenanceItem {
...
-        approveAccess?: NamedRef[];
+        approveAccess: NamedRef[];
     };
 }
 
 export interface GLASSMaintenancePaginatedObjects<T> extends PaginatedObjects<T> {
     rowIds: string[];
 }
+
+export function getUserModules(modules: GLASSModule[], user: User) {
```
Comment (tokland) [inline src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx:32]: [good practices] Explicit return type

File: src/webapp/reports/glass-admin/glass-admin-list/useDeleteFiles.tsx [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    const [isDeleteModalOpen, { enable: openDeleteModal, disable: closeDeleteModal }] = useBooleanState(false);
+
+    const deleteFiles = useCallback(
+        async (ids: string[]) => {
+            openDeleteModal();
+            compositionRoot.glassAdmin
+                .updateStatus(Namespaces.FILE_UPLOADS, "delete", ids)
+                ?.then(() => closeDeleteModal());
+            reload();
+        },
```
Comment (tokland) [inline src/webapp/reports/glass-admin/glass-admin-list/useDeleteFiles.tsx:21]: The function is defined as async, but no await is used. As it is, `reload` is executed without waiting fot updateState to finish, is that correct?

**Review by tokland — APPROVED** [review]
Thanks! Some minor comments, but code-wise, ready for merging

**Review by ifoche — CHANGES_REQUESTED** [review]
@deeonwuli a couple of comments:
- Please do not show anything until a module has been selected in the dropdown
- The ATC classification tab should only show up under AMC module selection

Could you please take care of this? 

Thank you!

**Review by ifoche — APPROVED** [review]
thanks @deeonwuli

**Pull request merged**